### PR TITLE
Add support for Attach to the kubelet.

### PR DESCRIPTION
**Pull request by brendandburns**
This is a pre-cursor to supporting `kubectl attach ...` and `kubectl run -it ...`

@dchen1107 @ncdc @vishh

### CODE REVIEW

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by derekwaynecarr** [comment]
@brendandburns - I always get confused between ```exec``` and ```attach``` but does this mean if/when we have ```kubectl attach ...``` that when I Ctrl-C, it will kill the container process in the same manner ```docker attach``` kills your container?

**Comment by brendandburns** [comment]
@derekwaynecarr yes, that's the idea.  `exec` is for running something *else* in the context of your container.  `attach` is for actually just taking over the existing running process in the container.

And so you could also hit ctrl-z to "background" it back to running in kubernetes.

**Comment by ncdc** [comment]
Docker supports ctrl p ctrl q to detach the attach session also

**Comment by derekwaynecarr** [comment]
I ask because on my Fedora 21 box I am unable to use Ctrl+q Ctrl+q for Docker, and the only way I have ever been able to detach without killing is if I attach with ```docker run --sig-proxy=false container```.

**Comment by derekwaynecarr** [comment]
Anyway, this PR LGTM for what its worth

**Comment by ncdc** [comment]
I can't review until next week but don't hold off merging on my account

**Comment by dchen1107** [comment]
LGTM for code. Also I can image how useful this feature could be since it provides a mechanism attaching stdin, stdout of a remote process (as a docker container). 

But I do have some concerns on introducing such feature to kubernetes based on today's design according the network traffic since all those traffic go through the master proxy. If many users try to attach containers, and send over large data through our master proxy, any negative impact on master node?

**Comment by brendandburns** [comment]
@dchen1107 for now, each master node is in its own VM, so any damage is limited to that one VM.

Also, it doesn't seem to be any more dangerous than `exec` which we already support.

**Comment by dchen1107** [comment]
Shouldn't messing up master node's network due to lack of traffic control / management could cause inherent delays to other data packets for real cluster management, for example, node watching new workloads assigning to it? 

I agreed that it is not more dangerous than kubectl exec, just concerning another mechanism / channel to mess up master proxy / master node.

File: pkg/kubelet/server.go [hunk 1]
```diff
@@ -413,7 +449,22 @@ func (s *Server) handleExec(w http.ResponseWriter, req *http.Request) {
...
+	}
+	err = s.host.ExecInContainer(kubecontainer.GetPodFullName(pod), uid, container, u.Query()[api.ExecCommandParamm], stdinStream, stdoutStream, stderrStream, tty)
+	if err != nil {
+		msg := fmt.Sprintf("Error executing command in container: %v", err)
+		glog.Error(msg)
+		errorStream.Write([]byte(msg))
+	}
+}
 
+func (s *Server) createStreams(w http.ResponseWriter, req *http.Request) (io.Reader, io.WriteCloser, io.WriteCloser, io.WriteCloser, httpstream.Connection, bool, bool) {
```
Comment (vishh) [inline pkg/kubelet/server.go]: That's a long list of return types :)

**Comment by vishh** [comment]
LGTM. I share's Dawn's concern with resource usage on API server and kubelet though.

**Comment by brendandburns** [comment]
Ok, can I get an lgtm label from @dchen1107 if we're ok to merge this (even with concerns about network usage)

(we can always throttle in the apiserver if we really start seeing problems...)

**Comment by dchen1107** [comment]
Ok, I marked this PR with lgtm since we generally have issue with network usage on master node without traffic control based on today's design. cc/ @bgrant0607, @davidopp on network resource management on master node as FYI.

**Comment by brendandburns** [comment]
I agree that I would rather see this handled with a global bandwidth shaping rule(s) on the master (e.g. https://wiki.debian.org/TrafficControl) rather than enabling/disabling features.

**Pull request merged**